pydantic/main.py

```diff
@@ -866,12 +904,41 @@ def __eq__(self, other: Any) -> bool:
             self_type = self.__pydantic_generic_metadata__['origin'] or self.__class__
             other_type = other.__pydantic_generic_metadata__['origin'] or other.__class__
 
-            return (
+            # Perform common checks first
+            if not (
                 self_type == other_type
-                and self.__dict__ == other.__dict__
                 and self.__pydantic_private__ == other.__pydantic_private__
                 and self.__pydantic_extra__ == other.__pydantic_extra__
-            )
+            ):
+                return False
+
+            # Fix GH-7444 by comparing only pydantic fields
+            # We provide a fast-path for performance: __dict__ comparison is *much* faster
+            # See tests/benchmarks/test_basemodel_eq_performances.py and GH-7825 for benchmarks
+            if self.__dict__ == other.__dict__:
+                # If the check above passes, then pydantic fields are equal, we can return early
+                return True
+            else:
+                # Else, we need to perform a more detailed, costlier comparison
+                # We use operator.itemgetter because it is much faster than dict comprehensions
+                # NOTE: Contratry to standard python class and instances, when the Model class has
+                #       attribute default values and the model instance doesn't has a corresponding
+                #       attribute, accessing the missing attribute raises an error in
+                #       __getattr__ instance of returning the class attribute
+                #      Thus, using operator.itemgetter() instead of operator.attrgetter() is valid
+                model_fields = type(self).model_fields.keys()
+                getter = operator.itemgetter(*model_fields) if model_fields else lambda _: _SENTINEL
+                try:
+                    return getter(self.__dict__) == getter(other.__dict__)
```

Comment:
Did you look at checking `self.__dict__.keys() != other.__dict__.keys()` as suggested in https://github.com/pydantic/pydantic/issues/7444#issuecomment-1753091873?

Reply:
What about the following dict:
- `self.__dict__ == {"a_field": 0, "_not_a_field": 1}`
- `other.__dict__ == {"a_field": 0, "_not_a_field": 99}`
?

They have the same keys, but we still want to trigger the detailed comparison with itemgetter.

Reply:
It could work with a more detailed check on keys taking into account the model fields, something like
`not (self.__dict__.keys() == other.__dict__.keys() == cls.model_fields.keys())`.

I'll need to think about this, but this is a good idea IMHO

Reply:
Added a fast-path for standard-case (i.e. no additional keys in __dict__ to ignore) unequal items. Previously, if the two __dict__ compared unequal, we would always trigger the slow __dict__ filtering code. This no longer happens when there are no keys to ignore.